src/include/distributed/master_metadata_utility.h

```diff
@@ -69,14 +69,32 @@ typedef struct ShardInterval
 } ShardInterval;
 
 
-/* In-memory representation of a tuple in pg_dist_shard_placement. */
+/* In-memory representation of a tuple in pg_dist_placement. */
+typedef struct GroupShardPlacement
+{
+	CitusNode type;
+	uint64 placementId;     /* sequence that implies this placement creation order */
+	uint64 shardId;
+	uint64 shardLength;
+	RelayFileState shardState;
+	uint32 groupId;
+} GroupShardPlacement;
+
+
+/* A GroupShardPlacement which has had some extra data resolved */
 typedef struct ShardPlacement
 {
+	/*
+	 * careful, the rest of the code assumes this exactly matches GroupShardPlacement
```

Comment:
> But if you actually want to express this, it should probably just embed the other struct?

I thought about that but that requires an even bigger change, have to go change every place those fields are accessed. It's probably the right answer though, I'll push another commit